sdk/core/azure-core/src/main/java/com/azure/core/http/policy/RedirectPolicy.java

```diff
@@ -0,0 +1,103 @@
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpMethod;
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.util.CoreUtils;
+import reactor.core.publisher.Mono;
+
+import java.net.HttpURLConnection;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * A HttpPipeline policy that redirects when an HTTP Redirect is received as response.
+ */
+public final class RedirectPolicy implements HttpPipelinePolicy {
+
+    private static final int PERMANENT_REDIRECT_STATUS_CODE = 308;
+
+    // Based on Stamp specific redirects design doc
+    private static final int MAX_REDIRECT_ATTEMPTS = 10;
+    private final RedirectStrategy redirectStrategy;
+
+    private Set<String> attemptedRedirectUrls = new HashSet<>();
+    private String redirectedEndpointUrl;
+
+    /**
+     * Creates {@link RedirectPolicy} with default {@link MaxAttemptRedirectStrategy} as {@link RedirectStrategy} and
+     * use the provided {@code statusCode} to determine if this request should be redirected
+     * and MAX_REDIRECT_ATTEMPTS for the try count.
+     */
+    public RedirectPolicy() {
+        this(new MaxAttemptRedirectStrategy(MAX_REDIRECT_ATTEMPTS));
+    }
+
+    /**
+     * Creates {@link RedirectPolicy} with default {@link MaxAttemptRedirectStrategy} as {@link RedirectStrategy} and
+     * use the provided {@code statusCode} to determine if this request should be redirected.
+     *
+     * @param redirectStrategy The {@link RedirectStrategy} used for redirection.
+     * @throws NullPointerException When {@code statusCode} is null.
+     */
+    public RedirectPolicy(RedirectStrategy redirectStrategy) {
+        this.redirectStrategy = Objects.requireNonNull(redirectStrategy, "'redirectStrategy' cannot be null.");
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        return attemptRedirect(context, next, context.getHttpRequest(), 1);
+    }
+
+    /**
+     * Function to process through the HTTP Response received in the pipeline
+     * and redirect sending the request with new redirect url.
+     */
+    private Mono<HttpResponse> attemptRedirect(final HttpPipelineCallContext context,
+                                               final HttpPipelineNextPolicy next,
+                                               final HttpRequest originalHttpRequest,
+                                               final int redirectAttempt) {
+        // make sure the context is not modified during redirect, except for the URL
+        context.setHttpRequest(originalHttpRequest.copy());
+        if (this.redirectedEndpointUrl != null) {
+            context.getHttpRequest().setUrl(this.redirectedEndpointUrl);
+        }
+        return next.clone().process()
```

Comment:
Let's add verbose level logging when a redirect is performed, this will help generate breadcrumbs when troubleshooting potential issues